admob/src/common/full_screen_ad_event_listener.cc

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2016 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "admob/src/common/full_screen_ad_event_listener.h"
+
+#include "app/src/mutex.h"
+
+namespace firebase {
+namespace admob {
+namespace internal {
+
+void FullScreenAdEventListener::SetFullScreenContentListener(
+    FullScreenContentListener* listener) {
+  MutexLock lock(listener_mutex_);
+  full_screen_content_listener_ = listener;
+}
+
+void FullScreenAdEventListener::SetPaidEventListener(
+    PaidEventListener* listener) {
+  MutexLock lock(listener_mutex_);
+  paid_event_listener_ = listener;
+}
+
+void FullScreenAdEventListener::NotifyListenerOfAdClickedFullScreenContent() {
+  MutexLock lock(listener_mutex_);
+  if (full_screen_content_listener_ != nullptr) {
+    full_screen_content_listener_->OnAdClicked();
```

Comment:
Just a question: is it safe to keep the mutex locked the whole time the user's callback is running? This could potentially be a long time. Typically if there's any way to avoid locking a mutex while a user callback runs we try not to lock it. Not really sure how you'd safely do this here though, to avoid the listener itself being deleted, so maybe never mind,.

Reply:
Yes, I see you're point.

I feel like tackling it this way:  If this mutex lock is a problem then we'll have safe customer reports saying that it's in their way.  But if we take the mutex out there, then we _could_ have crazy crash reports due to corrupted memory somewhere later in the runtime. For now I'd like to play it safe and open up to dangers later, but only if we need to.  WDYT?